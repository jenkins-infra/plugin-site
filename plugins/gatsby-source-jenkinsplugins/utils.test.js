/**
 * @jest-environment node
 */
const fs = require('fs');
const path = require('path');
const utils = require('./utils');
const nock = require('nock');
nock.disableNetConnect();

process.env.GET_CONTENT = true;

const readText = async (fileName) => {const buffer = await fs.promises.readFile(path.join(__dirname, '__mocks__', fileName)); return buffer.toString();};

describe('Utils', () => {
    let _reporter;
    afterEach(() => {
        nock.cleanAll();
    });
    beforeEach(async () => {
        _reporter = {
            panic: (...args) => { throw args[0]; },
            activityTimer: () => {
                return {
                    start: jest.fn(),
                    end: jest.fn(),
                };
            }
        };
    });
    it('Fix GitHub URL: submodule gets expanded', () => {
        expect(utils.fixGitHubUrl('https://github.com/jenkinsci/blueocean-plugin/blueocean-bitbucket-pipeline', 'master'))
            .toBe('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline');
    });
    it('Fix GitHub URL: expanded stays expanded', () => {
        expect(utils.fixGitHubUrl('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline', 'master'))
            .toBe('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline');
    });
    it('Fix GitHub URL: no submodule, keep short', () => {
        expect(utils.fixGitHubUrl('https://github.com/jenkinsci/junit-plugin', ''))
            .toBe('https://github.com/jenkinsci/junit-plugin');
    });
    it('Get plugin data for a wiki based plugin', async () => {
        nock('https://updates.jenkins.io')
            .get('/update-center.actual.json')
            .reply(200, JSON.parse(await readText('update-center.actual.json')));
        nock('https://updates.jenkins.io')
            .get('/current/plugin-documentation-urls.json')
            .reply(200, '{"ios-device-connector":{"url":"https://wiki.jenkins-ci.org/display/JENKINS/iOS+Device+Connector+Plugin"}}');
        nock('https://www.jenkins.io')
            .get('/doc/pipeline/steps/contents.json')
            .reply(200, []);
        nock('https://raw.githubusercontent.com:443')
            .get('/jenkinsci/bom/master/bom-latest/pom.xml')
            .reply(200, '');
        nock('https://raw.githubusercontent.com:443')
            .get('/jenkinsci/jenkins/master/core/src/main/resources/jenkins/split-plugins.txt')
            .reply(200, await readText('split-plugins.txt'));
        nock('https://plugins.jenkins.io').persist()
            .get(/\/api\/plugin\/.*/)
            .reply(200, '{"wiki":{"content": ""}}');
        nock('https://wiki.jenkins.io')
            .get('/rest/api/content?expand=body.view&title=iOS+Device+Connector+Plugin')
            .replyWithFile(200, path.join(__dirname, '__mocks__', 'wiki.jenkins.io.io-device-connector-plugin.json'), {'Content-Type': 'application/json'});

        const createNode = jest.fn().mockResolvedValue();
        const firstReleases = {'ios-device-connector': new Date(0)};
        const labelToCategory = {'ios': 'languagesPlatforms', 'builder': 'buildManagement'};
        const stats = {
            'core': {'installations': [{total: 300000}, {total: 300000}]},
            'ios-device-connector': {'currentInstalls': 269, 'installations': [
                {'timestamp': 1590984000000, 'total': 349},
                {'timestamp': 1593576000000, 'total': 348},
                {'timestamp': 1596254400000, 'total': 341},
                {'timestamp': 1598932800000, 'total': 347},
                {'timestamp': 1601524800000, 'total': 318},
                {'timestamp': 1604203200000, 'total': 325},
                {'timestamp': 1606798800000, 'total': 312},
                {'timestamp': 1609477200000, 'total': 301},
                {'timestamp': 1612155600000, 'total': 297},
                {'timestamp': 1614574800000, 'total': 303},
                {'timestamp': 1617249600000, 'total': 290},
                {'timestamp': 1619841600000, 'total': 275},
                {'timestamp': 1622520000000, 'total': 269}]}
        };
        await utils.fetchPluginData({createNode, reporter: _reporter, firstReleases, labelToCategory, stats});
        expect(createNode.mock.calls[0][0]).toMatchSnapshot();
    });
});
