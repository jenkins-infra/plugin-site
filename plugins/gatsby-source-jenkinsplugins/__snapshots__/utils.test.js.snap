// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils get plugin data for a wiki based plugin 1`] = `
Array [
  Object {
    "baseHref": "https://wiki.jenkins-ci.org/display/JENKINS/",
    "id": "ios-device-connector <<< JenkinsPluginWiki",
    "internal": Object {
      "content": "MISSING",
      "contentDigest": "10d1208b485425756fcc932229386b02",
      "mediaType": "text/pluginhtml",
      "type": "JenkinsPluginWiki",
    },
    "name": "ios-device-connector",
    "url": "https://wiki.jenkins-ci.org/display/JENKINS/iOS+Device+Connector+Plugin",
  },
  Object {
    "buildDate": 2012-10-10T00:00:00.000Z,
    "children": Array [],
    "defaultBranch": "master",
    "dependencies": Array [
      Object {
        "implied": true,
        "name": "external-monitor-job",
        "optional": false,
        "title": "External Monitor Job Type",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "ldap",
        "optional": false,
        "title": "LDAP",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "pam-auth",
        "optional": false,
        "title": "PAM Authentication",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "mailer",
        "optional": false,
        "title": "Mailer",
        "version": "1.2",
      },
      Object {
        "implied": true,
        "name": "matrix-auth",
        "optional": false,
        "title": "Matrix Authorization Strategy",
        "version": "1.0.2",
      },
      Object {
        "implied": true,
        "name": "windows-slaves",
        "optional": false,
        "title": "WMI Windows Agents",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "antisamy-markup-formatter",
        "optional": false,
        "title": "OWASP Markup Formatter",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "matrix-project",
        "optional": false,
        "title": "Matrix Project",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "junit",
        "optional": false,
        "title": "JUnit",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "bouncycastle-api",
        "optional": false,
        "title": "bouncycastle API",
        "version": "2.16.0",
      },
      Object {
        "implied": true,
        "name": "command-launcher",
        "optional": false,
        "title": "Command Agent Launcher",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "jdk-tool",
        "optional": false,
        "title": "Oracle Java SE Development Kit Installer",
        "version": "1.0",
      },
      Object {
        "implied": true,
        "name": "jaxb",
        "optional": false,
        "title": "JAXB",
        "version": "2.3.0",
      },
      Object {
        "implied": true,
        "name": "trilead-api",
        "optional": false,
        "title": "Trilead API",
        "version": "1.0.4",
      },
      Object {
        "implied": true,
        "name": "sshd",
        "optional": false,
        "title": "SSHD",
        "version": "3.0.1",
      },
    ],
    "developers": Array [],
    "excerpt": "Talks to iOS devices connected to slaves and do stuff.",
    "firstRelease": "1970-01-01T00:00:00.000Z",
    "gav": "org.jenkins-ci.plugins:ios-device-connector:1.2",
    "hasBomEntry": false,
    "hasExtensions": false,
    "hasPipelineSteps": false,
    "id": "ios-device-connector <<< JenkinsPlugin",
    "internal": Object {
      "contentDigest": "1da1c7c8d56f5351f4d3c618759f7689",
      "type": "JenkinsPlugin",
    },
    "issueTrackers": Array [
      Object {
        "reportUrl": "https://www.jenkins.io/participate/report-issue/redirect/#17020",
        "type": "jira",
        "viewUrl": "https://issues.jenkins.io/issues/?jql=component=17020",
      },
    ],
    "labels": Array [
      "builder",
      "ios",
    ],
    "name": "ios-device-connector",
    "parent": null,
    "popularity": 275,
    "previousTimestamp": "2012-10-08T20:04:36.00Z",
    "previousVersion": "1.1",
    "releaseTimestamp": "2012-10-10T17:27:42.00Z",
    "requiredCore": "1.466",
    "scm": "https://github.com/jenkinsci/ios-device-connector-plugin",
    "securityWarnings": Array [],
    "sha1": "B2z2wTyA8l5uwz9HqD+xhG9Hjns=",
    "sha256": "Tray3+tyfE6xoTu1Rge9+INkcj5y7Bgw7nPm//saiYw=",
    "size": 1548862,
    "stats": Object {
      "currentInstalls": 269,
      "installations": Array [
        Object {
          "timestamp": 1590984000000,
          "total": 349,
        },
        Object {
          "timestamp": 1593576000000,
          "total": 348,
        },
        Object {
          "timestamp": 1596254400000,
          "total": 341,
        },
        Object {
          "timestamp": 1598932800000,
          "total": 347,
        },
        Object {
          "timestamp": 1601524800000,
          "total": 318,
        },
        Object {
          "timestamp": 1604203200000,
          "total": 325,
        },
        Object {
          "timestamp": 1606798800000,
          "total": 312,
        },
        Object {
          "timestamp": 1609477200000,
          "total": 301,
        },
        Object {
          "timestamp": 1612155600000,
          "total": 297,
        },
        Object {
          "timestamp": 1614574800000,
          "total": 303,
        },
        Object {
          "timestamp": 1617249600000,
          "total": 290,
        },
        Object {
          "timestamp": 1619841600000,
          "total": 275,
        },
        Object {
          "timestamp": 1622520000000,
          "total": 269,
        },
      ],
      "trend": -20,
    },
    "title": "iOS Device connector",
    "url": "https://updates.jenkins.io/download/plugins/ios-device-connector/1.2/ios-device-connector.hpi",
    "version": "1.2",
  },
]
`;

exports[`utils get plugin healthScore data 1`] = `
Array [
  Object {
    "children": Array [],
    "details": Array [
      Object {
        "coefficient": 1,
        "key": "update-center-plugin-publication",
        "value": 1,
      },
      Object {
        "coefficient": 1,
        "key": "security",
        "value": 1,
      },
      Object {
        "coefficient": 0.5,
        "key": "repository-configuration",
        "value": 0.7,
      },
      Object {
        "coefficient": 0.8,
        "key": "adoption",
        "value": 1,
      },
      Object {
        "coefficient": 0.8,
        "key": "deprecation",
        "value": 1,
      },
    ],
    "id": "aws-java-sdk-sns",
    "internal": Object {
      "contentDigest": "f789c60332b91b320525aea3639a7c73",
      "type": "JenkinsPluginHealthScore",
    },
    "parent": null,
    "timestamp": "2023-03-24T15:15:33.022244Z",
    "value": 96,
    "version": "1.12.406-370.v8f993c987059",
  },
]
`;
